### DISCUSSION TITLE
No way for client code to tell if a target handle is connectable

### DISCUSSION THREAD

**Discussion by snoozbuster** [bcakmakoglu/vue-flow#824] [Ideas]
I'm writing a custom connection line which allows users to drag edges directly onto a node and automatically have the edge connect to the input handle (in my project, there is only ever one). The code I've written for this looks like so:

```ts
const { connectionStartHandle, onConnectEnd, getNodes: nodes, addEdges } = useVueFlow();

onConnectEnd(() => {
  const insideRect = (rect: Rect) =>
    rect.x < props.targetX &&
    rect.x + rect.width > props.targetX &&
    rect.y < props.targetY &&
    rect.y + rect.height > props.targetY;

  const intersectingNodes = orderBy(
    unref(nodes).filter(node => insideRect(getRectOfNodes([node]))),
    'computedPosition.z',
    'desc'
  );
  const targetNode = intersectingNodes[0];
  if (
    targetNode &&
    unref(connectionStartHandle).nodeId !== targetNode.id &&
    targetNode.handleBounds.target?.length
    // I would like to put an additional check for "is target handle connectable" here
  ) {
    addEdges([
      {
        id: getEdgeId(
          unref(connectionStartHandle).nodeId,
          unref(connectionStartHandle).handleId as any
        ),
        type: 'step',
        source: unref(connectionStartHandle).nodeId,
        sourceHandle: unref(connectionStartHandle).handleId,
        target: targetNode.id,
        targetHandle: 'input',
      },
    ]);
  }
});
```

In the `if` statement where I decide if I should make the connection, all I can really do is check if the handle exists. I don't have any way to actually check if the handle is connectable or not. Ideally I would like to re-use something like [`isValidHandle`](https://github.com/bcakmakoglu/vue-flow/blob/2543e1623048a43eac6ebf3fa11575c21a40f99f/packages/core/src/utils/handle.ts#L69), but that function isn't really useable as-is because it requires you to give it a mouse event and then derives the handle from that event. I already have the handle (or at least, its ID) and just want to check if the connection I'm about to make is valid, according to the handle props I've already passed to `<Handle />`.

**Comment by bcakmakoglu** [comment]
Connectable handles have a class name called `connectable` so you can just use something like 
```ts
document.querySelector(`[data-handleid="${handleId}"].connectable`)
``` 

to find a handle element that's connectable.

Or if you just need to validate if that element is connectable this would do it as well
```ts
document.querySelector(`[data-handleid="${handleId}"]`).classList.contains('connectable')
```

**Reply by bcakmakoglu** [reply]
With the next minor release I'll add a property `connectionTargetHandle` to the store as well, that way finding out some info about the target handle will be easier.
Right now, the handle result is stored in `connectionStartHandle.result` 

This contains the following type
```ts
interface ValidHandleResult {
  handleId: string | null
  handleDomNode: Element | null
  isValid: boolean
  connection: Connection
}
```

I might remove the result property in the next major again, so be aware of that. `connectionTargetHandle` will stay though.

**Reply by snoozbuster** [reply]
it's not just as simple as checking for the `connectable` class, right? we also need to run the source/target validation functions, if any, and make sure that we're checking the connection mode... most of the things that `isValidHandle` are doing still seem very relevant for something like this. 

(Also, I don't know if `connectionTargetHandle` would work for this case, since there would be no target handle in onConnectEnd - the whole point of my implementation there is to treat the entire node as the target handle, even if it wasn't within the `connectionRadius` and thus snapping automatically to the target)

**Reply by bcakmakoglu** [reply]
If you're not using a target handle why not just validate if the target node is ok?
You could just call your own validation for that, as you should be able to find out whatever target node you were on when `onConnectEnd` is triggered.

If you're trying to somehow replicate the `isValid` flag of a connection result, the code could just be re-used
```ts
// in strict mode we don't allow target to target or source to source connections
const isValid =
      handleToCheck.classList.contains('connectable') &&
      (connectionMode === ConnectionMode.Strict
        ? (isTarget && handle.type === 'source') || (!isTarget && handle.type === 'target')
        : handleNodeId !== fromNodeId || handleId !== fromHandleId)
```

Another option is to just use a target node that wraps your whole custom node so you actually don't work around the missing target node.

Other than that I don't see what else the library could even provide.

**Reply by snoozbuster** [reply]
>If you're not using a target handle why not just validate if the target node is ok?
You could just call your own validation for that

My plan was to re-use the validation framework of vue-flow to handle connection validation, passing `is-valid-connection` to the handle and such. I guess I don't strictly need to, but then I would need to replicate a lot of the guards and checks that the library has internally, when I would rather just trust the library.

>Another option is to just use a target node that wraps your whole custom node so you actually don't work around the missing target node.

I assume you meant "use a target handle that wraps your custom node"? Using a handle that wraps the entire node is somewhat intriguing, although I'm not really sure how I would do that offhand. I also need to make sure that connections can't be initiated from that handle. If I could do that though I wouldn't need the custom connection line or to reproduce any of the logic that the library needs. Can you elaborate a little more on how wrapping the whole node in a handle would work?

**Reply by bcakmakoglu** [reply]
I would think what you're trying to achieve is similar to the [easy connect example](https://reactflow.dev/docs/examples/nodes/easy-connect/) of React Flow?
What's done in that example should also work for Vue Flow I'd assume 😄

**Reply by snoozbuster** [reply]
yeah, I can adapt that. Thanks for the tip

**Reply by snoozbuster** [reply]
Actually - what they do in that example doesn't work here. All they do is make the handle absolutely positioned and cover 100% of the node, but vue-flow doesn't really care much about the exact handle dimensions. [`getClosestHandle`](https://github.com/bcakmakoglu/vue-flow/blob/707aaef5350a308c979f63a85d575491df80caf9/packages/core/src/utils/handle.ts#L55) only considers the point-to-point distance to the handle's "position" - which [for a 100% height/width, `Position.Top`](https://github.com/bcakmakoglu/vue-flow/blob/707aaef5350a308c979f63a85d575491df80caf9/packages/core/src/utils/edge.ts#L12) handle, is basically exactly the same as a 15x15 Position.Top handle. At no point does `getClosestHandle` consider the actual size of the handle or if the mouse is inside its bounding box - so unfortunately, this trick doesn't work 😞 

Based on the z-index manipulation they do in that example, I imagine react-flow actually uses the pointer events or bounding box of the handle element itself in order to decide where to place a connection

**Reply by bcakmakoglu** [reply]
Have you actually tried it though?
The implementation of `getClosestHandle` is the same in React and Vue Flow. No difference.

**Reply by bcakmakoglu** [reply]
[Here's a sandbox that reproduces the React Flow example.](https://codesandbox.io/p/sandbox/reverent-cartwright-49yve2?file=%2Fsrc%2FCustomConnectionLine.vue&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A28%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A28%7D%5D)

Though it does use the latest `1.18.0` version.

**Reply by snoozbuster** [reply]
weird... I did try it but I wasn't seeing any effect in my project. I'll give it another shot.

**Reply by bcakmakoglu** [reply]
I was running into trouble with 1.17.6 as well for some reason I’m not 100% sure of 😅 but with ^1.18.0 it works fine

**Reply by snoozbuster** [reply]
Thanks @bcakmakoglu. my previous implementation had a `visibility: hidden` style on the input handle to hide it. It seems like for whatever reason this causes it to not work. I was able to get it working though!
